single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -756,3 +796,15 @@ def _initialise_replica_set(self):
     def is_removing_last_replica(self) -> bool:
         """Returns True if the last replica (juju unit) is getting removed."""
         return self.state.planned_units == 0 and len(self.state.peers_units) == 0
+
+    def assert_proceed_on_broken_event(self, relation: Relation):
+        """Runs some checks on broken relation event."""
+        if not self.state.has_departed_run(relation.id):
+            raise DeferrableFailedHookChecksError(
+                "must wait for relation departed hook to decide if relation should be removed"
+            )
+
+        if self.state.is_scaling_down(relation.id):
+            raise NonDeferrableFailedHookChecksError(
```

Comment:
As you have noticed, I decided to go for a assert/raise pattern: if all checks are valid, we do nothing, if the check fails we raise and it's caught by the event handler which reacts according to the error raise: Non deferrable means we just return, deferrable means we log and defer, and uncaught exceptions will put into error (because it still needs to happen in some cases).
This makes a simple and straightforward pattern and avoid mixing booleans and error catching :)